### PR TITLE
$redirect priority

### LINKED ISSUE #4201: Support $redirect priorities
There are filters that specify priorities for $redirects. We should support that as currently the filters are break due to the priority being considered as a part of the name of the redirection target.

Few examples:
```
||g.doubleclick.net/tag/js/gpt.js$script,redirect=googletagservices_gpt.js:5,domain=euronews.com,important
||googlesyndication.com/pagead/js/adsbygoogle.js$script,redirect=noop.js:10,domain=1shortlink.com
||bannertrack.net^$script,3p,redirect=nobab2.js:10
*/fuckadblock-$script,redirect=fuckadblock.js-3.2.0:5
```

Should be implemented after #4200

### PR DESCRIPTION
**Pull request by chrmod**
fix #4201

### CODE REVIEW

File: packages/adblocker/test/engine/engine.test.ts [hunk 1]
```diff
@@ -445,6 +485,51 @@ $csp=baz,domain=bar.com
...
+        'foo.js',
+      ).match(request);
+      expect(redirect).to.eql({
+        body: 'foo.js',
+        contentType: 'application/javascript',
+        dataUrl: 'data:application/javascript;base64,Zm9vLmpz',
+      });
+    });
+
+    it('redirect-rule with with no priority wins', () => {
```
Comment (philipp-classen) [inline packages/adblocker/test/engine/engine.test.ts:500]: https://github.com/gorhill/uBlock/wiki/Static-filter-syntax#redirect

> By default, redirect directives have an implicit priority of 0. Filter authors can declare explicitness by appending :[integer] (negative values are also supported) to the redirect= option token.

I read it as `a.js` (priority 10) should win here.

File: packages/adblocker/src/filters/network.ts [hunk 2]
```diff
@@ -1368,6 +1368,14 @@ export default class NetworkFilter implements IFilter {
     return this.optionValue ?? '';
   }
 
+  public getRedirectResource(): string {
+    return this.getRedirect().split(':')[0];
+  }
+
+  public getRedirectPriority(): number {
```
Comment (philipp-classen) [inline packages/adblocker/src/filters/network.ts:1375]: Don't know if it is used in a performance critical path. If so, it could be optimized.

Comment (seia-soto) [reply packages/adblocker/src/filters/network.ts:1375]: I created a branch to handle this: https://github.com/ghostery/adblocker/pull/4219

**Review by seia-soto — APPROVED** [review]

File: packages/adblocker/src/filters/network.ts (hunk 2 above)
Comment (chrmod) [reply packages/adblocker/src/filters/network.ts:1375]: it's not a critical path. those are called after filters are already matched

File: packages/adblocker/src/filters/network.ts [hunk 3]
```diff
@@ -588,7 +588,7 @@ export default class NetworkFilter implements IFilter {
             }
 
             // Ignore this filter if no redirection resource is specified
-            if (value.length === 0) {
+            if (value.length === 0 || value.startsWith(':') || value.endsWith(':')) {
```
Comment (remusao) [inline packages/adblocker/src/filters/network.ts:591]: Do you have some examples where the value starts or ends with `:`?

**Review by remusao — APPROVED** [review]
Looks good once this is also merged: https://github.com/ghostery/adblocker/pull/4219

**Review by philipp-classen — APPROVED** [review]

**Pull request merged**